File: android/app/src/main/java/com/transcribecare/app/service/AudioRecorderService.kt
package com.transcribecare.app.service

import android.content.Context
import android.media.MediaRecorder
import android.os.Build
import android.os.StatFs
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Wrapper around Android's MediaRecorder that provides AAC audio capture
 * with file path management and graceful error handling.
 *
 * Recordings are stored in the app's external files directory (or internal files
 * directory as fallback) using timestamped filenames for uniqueness.
 *
 * @param onError Callback invoked with a user-facing error message when recording
 *               encounters an issue (microphone in use, storage full, init failure).
 */
class AudioRecorderService(
    private val onError: (message: String) -> Unit
) {

    private var mediaRecorder: MediaRecorder? = null
    private var currentFilePath: String? = null
    private var isRecording: Boolean = false

    companion object {
        private const val AUDIO_FILE_PREFIX = "recording_"
        private const val AUDIO_FILE_EXTENSION = ".m4a"
        private const val RECORDINGS_DIR = "recordings"
        private const val MIN_STORAGE_BYTES = 10 * 1024 * 1024L // 10 MB minimum
    }

    /**
     * Starts audio recording in AAC format (MPEG-4 container).
     *
     * Creates a unique file in the app's recordings directory and begins capture.
     * Returns the file path on success, or null if recording could not be started.
     *
     * @param context Application context used to resolve the recordings directory.
     * @return The absolute file path of the recording, or null on failure.
     */
    fun startRecording(context: Context): String? {
        if (isRecording) {
            return currentFilePath
        }

        val outputFile = createOutputFile(context) ?: return null

        if (!hasAvailableStorage(outputFile)) {
            onError("Not enough storage space to record audio.")
            return null
        }

        return try {
            val recorder = createMediaRecorder(context, outputFile.absolutePath)
            recorder.prepare()
            recorder.start()

            mediaRecorder = recorder
            currentFilePath = outputFile.absolutePath
            isRecording = true

            currentFilePath
        } catch (e: IOException) {
            cleanup()
            outputFile.delete()
            onError("Microphone is in use by another app.")
            null
        } catch (e: IllegalStateException) {
            cleanup()
            outputFile.delete()
            onError("Failed to initialize audio recorder. Please try again.")
            null
        } catch (e: RuntimeException) {
            cleanup()
            outputFile.delete()
            onError("Failed to start recording. Please try again.")
            null
        }
    }

    /**
     * Stops the current recording and finalizes the audio file.
     *
     * @return The file path of the completed recording, or null if no recording was active.
     */
    fun stopRecording(): String? {
        if (!isRecording) {
            return null
        }

        val filePath = currentFilePath

        try {
            mediaRecorder?.stop()
        } catch (e: RuntimeException) {
            // stop() can throw if no audio data was recorded
            filePath?.let { File(it).delete() }
            cleanup()
            onError("Recording failed. No audio data was captured.")
            return null
        }

        cleanup()
        return filePath
    }

    /**
     * Releases all resources held by the MediaRecorder.
     * Should be called when the service is no longer needed.
     */
    fun release() {
        if (isRecording) {
            try {
                mediaRecorder?.stop()
            } catch (_: RuntimeException) {
                // Ignore — we're cleaning up regardless
            }
        }
        cleanup()
    }

    /**
     * Returns the file path of the current or most recent recording.
     */
    fun getCurrentFilePath(): String? = currentFilePath

    /**
     * Returns whether a recording is currently in progress.
     */
    fun isCurrentlyRecording(): Boolean = isRecording

    @Suppress("DEPRECATION")
    private fun createMediaRecorder(context: Context, outputPath: String): MediaRecorder {
        val recorder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            MediaRecorder(context)
        } else {
            MediaRecorder()
        }

        recorder.apply {
            setAudioSource(MediaRecorder.AudioSource.VOICE_RECOGNITION)
            setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            setAudioEncodingBitRate(128000)
            setAudioSamplingRate(44100)
            setOutputFile(outputPath)
            setOnErrorListener { _, what, _ ->
                val message = when (what) {
                    MediaRecorder.MEDIA_RECORDER_ERROR_UNKNOWN ->
                        "An unexpected recording error occurred."
                    MediaRecorder.MEDIA_ERROR_SERVER_DIED ->
                        "Recording service stopped unexpectedly."
                    else ->
                        "Recording error occurred."
                }
                onError(message)
                cleanup()
            }
        }

        return recorder
    }

    private fun createOutputFile(context: Context): File? {
        val recordingsDir = getRecordingsDirectory(context)
        if (recordingsDir == null) {
            onError("Unable to access storage for recording.")
            return null
        }

        val timestamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date())
        val fileName = "$AUDIO_FILE_PREFIX$timestamp$AUDIO_FILE_EXTENSION"

        return File(recordingsDir, fileName)
    }

    private fun getRecordingsDirectory(context: Context): File? {
        // Prefer external files directory (accessible but app-private)
        val externalDir = context.getExternalFilesDir(RECORDINGS_DIR)
        if (externalDir != null && (externalDir.exists() || externalDir.mkdirs())) {
            return externalDir
        }

        // Fallback to internal files directory
        val internalDir = File(context.filesDir, RECORDINGS_DIR)
        if (internalDir.exists() || internalDir.mkdirs()) {
            return internalDir
        }

        return null
    }

    private fun hasAvailableStorage(outputFile: File): Boolean {
        return try {
            val parentDir = outputFile.parentFile ?: return false
            val stat = StatFs(parentDir.absolutePath)
            stat.availableBytes > MIN_STORAGE_BYTES
        } catch (_: IllegalArgumentException) {
            false
        }
    }

    private fun cleanup() {
        try {
            mediaRecorder?.reset()
            mediaRecorder?.release()
        } catch (_: Exception) {
            // Ignore cleanup errors
        }
        mediaRecorder = null
        isRecording = false
    }
}

File: ios/TranscribeCare/Service/AudioRecorderService.swift
import Foundation
import AVFoundation

/// Wrapper around AVAudioRecorder that provides M4A/AAC audio capture
/// with file path management and graceful error handling.
///
/// Recordings are stored in the app's documents directory using timestamped
/// filenames for uniqueness.
///
/// - Requirements: 5.1, 5.2, 5.6
@Observable
class AudioRecorderService {

    // MARK: - Callbacks

    /// Called with a user-facing error message when recording encounters a problem.
    var onError: ((String) -> Void)?

    // MARK: - Public State

    /// Whether a recording is currently in progress.
    private(set) var isRecording: Bool = false

    /// The file URL of the current or most recent recording.
    private(set) var currentRecordingURL: URL?

    // MARK: - Private Properties

    private var audioRecorder: AVAudioRecorder?

    // MARK: - Constants

    private static let recordingsDirectory = "Recordings"
    private static let filePrefix = "recording_"
    private static let fileExtension = "m4a"
    private static let minimumStorageBytes: UInt64 = 10 * 1024 * 1024 // 10 MB

    // MARK: - Audio Settings

    /// AAC encoding settings for M4A audio capture.
    private var recordingSettings: [String: Any] {
        [
            AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
            AVSampleRateKey: 44100.0,
            AVNumberOfChannelsKey: 1,
            AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue,
            AVEncoderBitRateKey: 128000
        ]
    }

    // MARK: - Public Methods

    /// Starts audio recording in M4A/AAC format.
    ///
    /// Configures the audio session for recording, creates a unique output file,
    /// and begins capture. Returns the file URL on success, or nil on failure.
    ///
    /// - Returns: The URL of the recording file, or nil if recording could not start.
    func startRecording() -> URL? {
        if isRecording {
            return currentRecordingURL
        }

        // Configure audio session
        guard configureAudioSession() else {
            return nil
        }

        // Create output file URL
        guard let outputURL = createOutputFileURL() else {
            return nil
        }

        // Check available storage
        guard hasAvailableStorage() else {
            onError?("Not enough storage space to record audio.")
            return nil
        }

        // Create and start the recorder
        do {
            let recorder = try AVAudioRecorder(url: outputURL, settings: recordingSettings)
            recorder.prepareToRecord()

            guard recorder.record() else {
                onError?("Failed to start recording. Please try again.")
                return nil
            }

            audioRecorder = recorder
            currentRecordingURL = outputURL
            isRecording = true

            return outputURL
        } catch let error as NSError {
            handleRecorderInitError(error, fileURL: outputURL)
            return nil
        }
    }

    /// Stops the current recording and finalizes the audio file.
    ///
    /// - Returns: The URL of the completed recording, or nil if no recording was active.
    func stopRecording() -> URL? {
        guard isRecording, let recorder = audioRecorder else {
            return nil
        }

        recorder.stop()
        isRecording = false

        let fileURL = currentRecordingURL

        // Verify the file was actually written
        if let url = fileURL {
            let fileManager = FileManager.default
            if !fileManager.fileExists(atPath: url.path) {
                onError?("Recording failed. No audio data was captured.")
                cleanup()
                return nil
            }
        }

        // Deactivate audio session
        deactivateAudioSession()

        audioRecorder = nil
        return fileURL
    }

    /// Releases all resources held by the audio recorder.
    /// Should be called when the service is no longer needed.
    func release() {
        if isRecording {
            audioRecorder?.stop()
        }
        cleanup()
        deactivateAudioSession()
    }

    // MARK: - Private Methods

    /// Configures the AVAudioSession for recording.
    /// - Returns: `true` if configuration succeeded, `false` otherwise.
    private func configureAudioSession() -> Bool {
        let session = AVAudioSession.sharedInstance()

        do {
            try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker])
            try session.setActive(true, options: .notifyOthersOnDeactivation)
            return true
        } catch let error as NSError {
            if error.code == AVAudioSession.ErrorCode.insufficientPriority.rawValue ||
               error.domain == NSOSStatusErrorDomain {
                onError?("Microphone is in use by another app.")
            } else {
                onError?("Failed to configure audio session: \(error.localizedDescription)")
            }
            return false
        }
    }

    /// Deactivates the audio session after recording completes.
    private func deactivateAudioSession() {
        do {
            try AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
        } catch {
            // Non-critical — session will be deactivated when app backgrounds
        }
    }

    /// Creates a unique output file URL in the app's recordings directory.
    /// - Returns: A URL for the new recording file, or nil if the directory could not be created.
    private func createOutputFileURL() -> URL? {
        let fileManager = FileManager.default

        guard let documentsDirectory = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
            onError?("Unable to access storage for recording.")
            return nil
        }

        let recordingsDir = documentsDirectory.appendingPathComponent(Self.recordingsDirectory)

        // Create recordings directory if needed
        if !fileManager.fileExists(atPath: recordingsDir.path) {
            do {
                try fileManager.createDirectory(at: recordingsDir, withIntermediateDirectories: true)
            } catch {
                onError?("Unable to access storage for recording.")
                return nil
            }
        }

        // Generate unique filename using timestamp
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        let timestamp = formatter.string(from: Date())
        let fileName = "\(Self.filePrefix)\(timestamp).\(Self.fileExtension)"

        return recordingsDir.appendingPathComponent(fileName)
    }

    /// Checks whether sufficient storage is available for recording.
    /// - Returns: `true` if at least the minimum storage threshold is available.
    private func hasAvailableStorage() -> Bool {
        let fileManager = FileManager.default

        guard let documentsDirectory = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
            return false
        }

        do {
            let attributes = try fileManager.attributesOfFileSystem(forPath: documentsDirectory.path)
            if let freeSpace = attributes[.systemFreeSize] as? UInt64 {
                return freeSpace > Self.minimumStorageBytes
            }
            return false
        } catch {
            return false
        }
    }

    /// Handles errors from AVAudioRecorder initialization.
    private func handleRecorderInitError(_ error: NSError, fileURL: URL) {
        // Clean up the file that was created
        try? FileManager.default.removeItem(at: fileURL)

        if error.domain == NSOSStatusErrorDomain {
            // OS-level audio errors often indicate hardware issues
            onError?("Microphone is in use by another app.")
        } else {
            onError?("Failed to initialize audio recorder: \(error.localizedDescription)")
        }
    }

    /// Cleans up recorder resources and resets state.
    private func cleanup() {
        audioRecorder?.stop()
        audioRecorder = nil
        isRecording = false
    }
}
